### PR TITLE
[Bugfix] Jenkins user management

### LINKED ISSUE #3041: Jenkins breaks user services
### Describe the bug
Seems like #2959 (@ivanchimeno) introduces a number of issues.

#### To Reproduce
1. Start Artemis
- stack trace: [update-admin.log](https://github.com/ls1intum/Artemis/files/6156470/update-admin.log)
```
00 Bad Request: [{
artemis-be_1  | "servlet":"Stapler",
artemis-be_1  | "message":"Cannot delete self",
artemis-be_1  | "url":"/user/artemis/doDelete",
artemis-be_1  | "status":"400"
artemis-be_1  | }]
```

or

1. Start Artemis
2. Reset password
- stack trace [reset-pw.log](https://github.com/ls1intum/Artemis/files/6156463/reset-pw.log)

### Environment

 - OS: https://github.com/LDAP/docker-compose-artemis
 - Commit: a1da6936a11c33db9bc9763a89062773613f709a (aa7bd6610eb905c8c35b7dd9833b29d6d7c4a0fd)

### PR DESCRIPTION
**Pull request by LDAP**
### Checklist
- [ ] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context
Fixes #3041 

### Steps for Testing
See Issue

@krusche I added the user to the CI on registerUser too. Was there a reason to not add the user?

### CODE REVIEW

**Comment by ivanchimeno** [comment]
@LDAP, you were fast than me in creating the PR ;) I made some changes in the branch ` bugfix/jenkins-user-management/various-issues ` if you're interrested.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsUserManagementService.java [hunk 1]
```diff
@@ -41,6 +41,9 @@
     @Value("${artemis.continuous-integration.url}")
     private URL jenkinsServerUrl;
 
+    @Value("${artemis.continuous-integration.user}")
+    private String jenkinsAdminUsername;
```
Comment (ivanchimeno) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsUserManagementService.java:45]: Ah, that's a very nice way of solving the issue!

**Comment by LDAP** [comment]
> 
> 
> @LDAP, you were fast than me in creating the PR ;) I went over the code and couldn't find any issues. Should we rather guarantee that the groups are loaded as you suggested in the issue? This might prevent errors in the future.
> 
> I made some changes in the branch [`bugfix/jenkins-user-management/various-issues`](https://github.com/ls1intum/Artemis/compare/bugfix/jenkins-user-management/various-issues) which implement your suggestions. Let me know if we should incorporate that in this pr

That comes at the cost of an additional database request. Not sure what's better...

**Comment by ivanchimeno** [comment]
> > @LDAP, you were fast than me in creating the PR ;) I went over the code and couldn't find any issues. Should we rather guarantee that the groups are loaded as you suggested in the issue? This might prevent errors in the future.
> > I made some changes in the branch [`bugfix/jenkins-user-management/various-issues`](https://github.com/ls1intum/Artemis/compare/bugfix/jenkins-user-management/various-issues) which implement your suggestions. Let me know if we should incorporate that in this pr
> 
> That comes at the cost of an additional database request. Not sure what's better...

You're right. Stating that the groups must be loaded in the function java doc would be a good compromise.

**Comment by LDAP** [comment]
> 
> 
> > > @LDAP, you were fast than me in creating the PR ;) I went over the code and couldn't find any issues. Should we rather guarantee that the groups are loaded as you suggested in the issue? This might prevent errors in the future.
> > > I made some changes in the branch [`bugfix/jenkins-user-management/various-issues`](https://github.com/ls1intum/Artemis/compare/bugfix/jenkins-user-management/various-issues) which implement your suggestions. Let me know if we should incorporate that in this pr
> > 
> > 
> > That comes at the cost of an additional database request. Not sure what's better...
> 
> You're right. Stating that the groups must be loaded in the function java doc would be a good compromise.

In this case we should check every call to ensure that groups are loaded.

```
src\main\java\de\tum\in\www1\artemis\service\user\UserService.java:
  264:102  optionalCIUserManagementService.ifPresent(ciUserManagementService -> ciUserManagementService.deleteUser(existingUser));
  348:106  optionalCIUserManagementService.ifPresent(ciUserManagementService -> ciUserManagementService.deleteUser(user));

src\main\java\de\tum\in\www1\artemis\service\connectors\jenkins\JenkinsUserManagementService.java:
  149:13   deleteUser(user);
  175:9    deleteUser(oldUser);
src\main\java\de\tum\in\www1\artemis\service\user\UserCreationService.java:
  162:102  optionalCIUserManagementService.ifPresent(ciUserManagementService -> ciUserManagementService.createUser(user));

src\main\java\de\tum\in\www1\artemis\service\connectors\jenkins\JenkinsUserManagementService.java:
  146:13   createUser(user);
  150:13   createUser(user);
  177:9    createUser(user);
src\main\java\de\tum\in\www1\artemis\service\user\UserCreationService.java:
  190:106  optionalCIUserManagementService.ifPresent(ciUserManagementService -> ciUserManagementService.updateUser(user));

src\main\java\de\tum\in\www1\artemis\service\user\UserService.java:
  178:106  optionalCIUserManagementService.ifPresent(ciUserManagementService -> ciUserManagementService.updateUser(user));
  391:106  optionalCIUserManagementService.ifPresent(ciUserManagementService -> ciUserManagementService.updateUser(user));

src\main\java\de\tum\in\www1\artemis\service\connectors\jenkins\JenkinsUserManagementService.java:
  186:13   updateUser(user);
src\main\java\de\tum\in\www1\artemis\service\connectors\jenkins\JenkinsUserManagementService.java:
  183:13   updateUserLogin(oldLogin, user);
```

Now that I see how many references there are I may change my opinion. It would be indeed safer to load the groups directly in the Jenkins service.

**Comment by ivanchimeno** [comment]
> > > > @LDAP, you were fast than me in creating the PR ;) I went over the code and couldn't find any issues. Should we rather guarantee that the groups are loaded as you suggested in the issue? This might prevent errors in the future.
> > > > I made some changes in the branch [`bugfix/jenkins-user-management/various-issues`](https://github.com/ls1intum/Artemis/compare/bugfix/jenkins-user-management/various-issues) which implement your suggestions. Let me know if we should incorporate that in this pr
> > > 
> > > 
> > > That comes at the cost of an additional database request. Not sure what's better...
> > 
> > 
> > You're right. Stating that the groups must be loaded in the function java doc would be a good compromise.
> 
> In this case we should check every call to ensure that groups are loaded.
> 
> ```
> src\main\java\de\tum\in\www1\artemis\service\user\UserService.java:
>   264:102  optionalCIUserManagementService.ifPresent(ciUserManagementService -> ciUserManagementService.deleteUser(existingUser));
>   348:106  optionalCIUserManagementService.ifPresent(ciUserManagementService -> ciUserManagementService.deleteUser(user));
> 
> src\main\java\de\tum\in\www1\artemis\service\connectors\jenkins\JenkinsUserManagementService.java:
>   149:13   deleteUser(user);
>   175:9    deleteUser(oldUser);
> src\main\java\de\tum\in\www1\artemis\service\user\UserCreationService.java:
>   162:102  optionalCIUserManagementService.ifPresent(ciUserManagementService -> ciUserManagementService.createUser(user));
> 
> src\main\java\de\tum\in\www1\artemis\service\connectors\jenkins\JenkinsUserManagementService.java:
>   146:13   createUser(user);
>   150:13   createUser(user);
>   177:9    createUser(user);
> src\main\java\de\tum\in\www1\artemis\service\user\UserCreationService.java:
>   190:106  optionalCIUserManagementService.ifPresent(ciUserManagementService -> ciUserManagementService.updateUser(user));
> 
> src\main\java\de\tum\in\www1\artemis\service\user\UserService.java:
>   178:106  optionalCIUserManagementService.ifPresent(ciUserManagementService -> ciUserManagementService.updateUser(user));
>   391:106  optionalCIUserManagementService.ifPresent(ciUserManagementService -> ciUserManagementService.updateUser(user));
> 
> src\main\java\de\tum\in\www1\artemis\service\connectors\jenkins\JenkinsUserManagementService.java:
>   186:13   updateUser(user);
> src\main\java\de\tum\in\www1\artemis\service\connectors\jenkins\JenkinsUserManagementService.java:
>   183:13   updateUserLogin(oldLogin, user);
> ```
> 
> Now that I see how many references there are I may change my opinion. It would be indeed safer to load the groups directly in the Jenkins service.
> 
> On the other hand what happens if the User is first deleted in Artemis and then deleteUser is called. In this case a DB request would fail.

The `UserService` deletes users from the VCS and CI first, before deleting it from Artemis. So I think we're safe here.

**Comment by LDAP** [comment]
@ivanchimeno What do you think about this compromise? Groups are only fetched if not already initialized,

**Comment by ivanchimeno** [comment]
> @ivanchimeno What do you think about this compromise? Groups are only fetched if not already initialized,

That compromise looks good to me. The code also looks good overall. I couldn't test itlocally because the branch isn't part of the Artemis repo but I'll do that once it is.

**Comment by LDAP** [comment]
Closed in favor of #3052

**Pull request closed without merging**